src/interface/TWAnyAddress.cpp

```diff
@@ -0,0 +1,171 @@
+// Copyright © 2017-2020 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include <TrustWalletCore/TWAnyAddress.h>
+#include <TrustWalletCore/TWPublicKey.h>
+#include <TrezorCrypto/cash_addr.h>
+
+#include "../Bitcoin/Address.h"
+#include "../Bitcoin/CashAddress.h"
+#include "../Bitcoin/SegwitAddress.h"
+#include "../Cosmos/Address.h"
+#include "../Decred/Address.h"
+#include "../Kusama/Address.h"
+#include "../Polkadot/Address.h"
+#include "../Zcash/TAddress.h"
+#include "../Zilliqa/Address.h"
+#include "../Nano/Address.h"
+
+#include "../Coin.h"
+#include "../HexCoding.h"
+
+using namespace TW;
+
+struct TWAnyAddress {
+    TWString* address;
+    enum TWCoinType coin;
+};
+
+bool TWAnyAddressEqual(struct TWAnyAddress* _Nonnull lhs, struct TWAnyAddress* _Nonnull rhs) {
+    return TWStringEqual(lhs->address, rhs->address) && lhs->coin == rhs->coin;
+}
+
+bool TWAnyAddressIsValidString(TWString* _Nonnull string, enum TWCoinType coin) {
+    auto& address = *reinterpret_cast<const std::string*>(string);
+    return TW::validateAddress(coin, address);
+}
+
+struct TWAnyAddress* _Nullable TWAnyAddressCreateWithString(TWString* _Nonnull string,
+                                                            enum TWCoinType coin) {
+    if (!TWAnyAddressIsValidString(string, coin)) {
+        return nullptr;
+    }
+    return new TWAnyAddress{TWStringCreateWithUTF8Bytes(TWStringUTF8Bytes(string)), coin};
+}
+
+struct TWAnyAddress* _Nonnull TWAnyAddressCreateWithPublicKey(
+    struct TWPublicKey* _Nonnull publicKey, enum TWCoinType coin) {
+    auto address = TW::deriveAddress(coin, publicKey->impl);
+    return new TWAnyAddress{TWStringCreateWithUTF8Bytes(address.c_str()), coin};
+}
+
+void TWAnyAddressDelete(struct TWAnyAddress* _Nonnull address) {
+    TWStringDelete(address->address);
+    delete address;
+}
+
+TWString* _Nonnull TWAnyAddressDescription(struct TWAnyAddress* _Nonnull address) {
+    return TWStringCreateWithUTF8Bytes(TWStringUTF8Bytes(address->address));
+}
+
+TWData* _Nonnull TWAnyAddressData(struct TWAnyAddress* _Nonnull address) {
+    auto& string = *reinterpret_cast<const std::string*>(address->address);
+    Data data;
+    switch (address->coin) {
```

Comment:
It would be nicer design if there is a C++ AnyAddress, and TWAnyAddress delt with C-C++ only, but no logic in it (separation of concerns).

Reply:
I actually thought this, but we won't use `AnyAddress` in c++ code (each blockchain has its own Address class), this is mainly for C interface, if we have other Address implementation other than cpp, we can just tweak it here.